### ISSUE TITLE
Bug in implementation of crypt() produces invalid results

### ISSUE THREAD

**Issue by ossgeek314** [oerdnj/deb.sury.org#1935]
**Frequently asked questions**

- [X] I have read [Frequently Asked Questions](https://github.com/oerdnj/deb.sury.org/wiki/Frequently-Asked-Questions)
- [X] I have looked at the list of the existing issues (including closed issues) and searched if my issue has been already reported
- [X] I have tried to resolve the issue myself and will describe what I did in clear and consise manner

**Describe the bug**
When using the `crypt()` function from the php version in the sury repository, we get different results than using the same `crypt()` function in the vanilla install of ubuntu 22.0.4 LTS (php 8.1).  The sury version of `crypt()` returns a failure to create a hash of *0 but the standard install produces a valid hash. 

To reproduce run the following code on the standard install of ubuntu 22.0.4 php8.1 and then run the same code on the sury version (either 7.4 or 8.1).  You get these two different results:

```php
    $string = "SuperSecretString";
    $salt_str_21 = 'abcdefghijklmnopqrstu';
    $salt_str_22 = 'abcdefghijklmnopqrstuv';
    $salt_21 = '$2y$11$' . $salt_str_21 . '$';
    $salt_22 = '$2y$11$' . $salt_str_22 . '$';
    $hash_21 = crypt($string, $salt_21);
    $hash_22 = crypt($string, $salt_22);
    print "String: " . $string . "\n";
    print "Salt Length 21\n";
    print "Salt: " . $salt_21 . "\n";
    print "Hash: " . $hash_21 . "\n";
    print "Salt Length 22\n";
    print "Salt: " . $salt_22 . "\n";
    print "Hash: " . $hash_22 . "\n";
```
Here is the output from Ubuntu 22.0.4 default install php8.1
```
String: SuperSecretString
Salt Length 21
Salt: $2y$11$abcdefghijklmnopqrstu$
Hash: $2y$11$abcdefghijklmnopqrstu.S1LRjIpMYl4oz4Ll7serIicjsmLYrTu
Salt Length 22
Salt: $2y$11$abcdefghijklmnopqrstuv$
Hash: $2y$11$abcdefghijklmnopqrstuu8n9wyqiQGBrPrpZUWP/UkVpQyGR4lEW
```
Here is the output from Ubuntu 20.0.4 default install php7.4.30
```
String: SuperSecretString
Salt Length 21
Salt: $2y$11$abcdefghijklmnopqrstu$
Hash: $2y$11$abcdefghijklmnopqrstu.S1LRjIpMYl4oz4Ll7serIicjsmLYrTu
Salt Length 22
Salt: $2y$11$abcdefghijklmnopqrstuv$
Hash: $2y$11$abcdefghijklmnopqrstuu8n9wyqiQGBrPrpZUWP/UkVpQyGR4lEW

```
Here is the output from Ubuntu 22.04 with sury repo php7.4 installed
```
String: SuperSecretString
Salt Length 21
Salt: $2y$11$abcdefghijklmnopqrstu$
Hash: *0
Salt Length 22
Salt: $2y$11$abcdefghijklmnopqrstuv$
Hash: $2y$11$abcdefghijklmnopqrstuu8n9wyqiQGBrPrpZUWP/UkVpQyGR4lEW
``` 
Here is the output from Ubuntu 22.0.4 with sury repo php8.1 installed
```
String: SuperSecretString
Salt Length 21
Salt: $2y$11$abcdefghijklmnopqrstu$
Hash: *0
Salt Length 22
Salt: $2y$11$abcdefghijklmnopqrstuv$
Hash: $2y$11$abcdefghijklmnopqrstuu8n9wyqiQGBrPrpZUWP/UkVpQyGR4lEW
```
In all of them you notice with the salt length of 22 the last character is cut off from the hash.  This is the reason we use a salt length of 21 so that the full salt is retained in the hash output.  For some reason this fails with the sury repository.

**Your understanding of what is happening**
For some reason the salt string is being rejected in the sury repo versions of php whereas the default installs do not exhibit this error.

**What steps did you take to resolve issue yourself before reporting it here**
We tried to debug the crypt function as well as creating different simplified scripts to remove other variables.  The issue only shows up when using the sury repo version of php. (not indicated here but we also tested CentOS and Linux Mint default installs)

**Expected behavior**
The results from the `crypt()` function should be exactly the same between default install and sury repository versions of php

**Distribution (please complete the following information):**
 - OS: Ubuntu 22.0.4 and 20.04
 - Architecture: amd64
 - Repository: Ubuntu PPA vs packages.sury.org

**Package(s) (please complete the following information):**
for Ubuntu 22.0.4 php8.1
```
php:
  Installed: 2:8.1+92ubuntu1
  Candidate: 2:8.1+92ubuntu1
  Version table:
 *** 2:8.1+92ubuntu1 500
        500 http://us.archive.ubuntu.com/ubuntu jammy/main amd64 Packages
        100 /var/lib/dpkg/status
```
for Ubuntu 20.04 php7.4
```
php:
  Installed: 2:7.4+75
  Candidate: 2:7.4+75
  Version table:
 *** 2:7.4+75 500
        500 mirror://mirrors.ubuntu.com/mirrors.txt focal/main amd64 Packages
        500 mirror://mirrors.ubuntu.com/mirrors.txt focal/main i386 Packages
        100 /var/lib/dpkg/status

```
for Ubuntu 22.0.4 sury php8.1
```
php8.1:
  Installed: 8.1.16+repack-1+ubuntu22.04.1+deb.sury.org+1
  Candidate: 8.1.16+repack-1+ubuntu22.04.1+deb.sury.org+1
  Version table:
 *** 8.1.16+repack-1+ubuntu22.04.1+deb.sury.org+1 500
        500 https://ppa.launchpadcontent.net/ondrej/php/ubuntu jammy/main amd64 Packages
        500 https://ppa.launchpadcontent.net/ondrej/php/ubuntu jammy/main i386 Packages
        100 /var/lib/dpkg/status
     8.1.8-1+ubuntu22.04.1+deb.sury.org+1 500
        500 https://ppa.launchpadcontent.net/ondrej/php/ubuntu jammy/main amd64 Packages
        500 https://ppa.launchpadcontent.net/ondrej/php/ubuntu jammy/main i386 Packages
     8.1.2-1ubuntu2.10 500
        500 http://de.archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages
        500 http://de.archive.ubuntu.com/ubuntu jammy-updates/main i386 Packages
        500 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu jammy-security/main i386 Packages
     8.1.2-1ubuntu2 500
        500 http://de.archive.ubuntu.com/ubuntu jammy/main amd64 Packages
        500 http://de.archive.ubuntu.com/ubuntu jammy/main i386 Packages
```
for Ubuntu 22.0.4 sury php7.4
```
php7.4:
  Installed: 1:7.4.33-5+ubuntu22.04.1+deb.sury.org+1
  Candidate: 1:7.4.33-5+ubuntu22.04.1+deb.sury.org+1
  Version table:
 *** 1:7.4.33-5+ubuntu22.04.1+deb.sury.org+1 500
        500 https://ppa.launchpadcontent.net/ondrej/php/ubuntu jammy/main amd64 Packages
        500 https://ppa.launchpadcontent.net/ondrej/php/ubuntu jammy/main i386 Packages
        100 /var/lib/dpkg/status
```

**Additional context**
This code has been running for many years with sury and only recently has the `crypt()` function begun to produce different results in the sury repository

**Comment by oerdnj** [comment]
The Ubuntu versions are vulnerable to:
```
      . Fixed bug #81744 (Password_verify() always return true with some hash).
        (CVE-2023-0567). (Tim Düsterhus)
```

See https://github.com/php/php-src/commit/c840f71524067aa474c00c3eacfb83bd860bfc8a

Basically, stop using invalid BCrypt hashes.

**Comment by oerdnj** [comment]
Also:

> The results from the crypt() function should be exactly the same between default install and sury repository versions of php

Yes, I agree, but the bug is on Ubuntu's side...
